File: backend/orm/delay_taxonomy.py
"""
Justified-delay taxonomy (Cycle 2 of the reporting data-capture roadmap).

3-state model: WorkOrder.delay_classification is NULL (unclassified, default),
'justified', or 'unjustified'. Unclassified is the ABSENCE of a value — never
an enum member, never offered in UI selects. justified_delay_reason is stored
only when classification == 'justified'.

Sibling of backend/orm/downtime_taxonomy.py (Cycle 1) — same conventions.
Spec: docs/superpowers/specs/2026-08-04-justified-delay-flag-design.md
"""

from enum import Enum


class DelayClassificationEnum(str, Enum):
    """Late-order delay classification (NULL = unclassified)."""

    JUSTIFIED = "justified"
    UNJUSTIFIED = "unjustified"


class JustifiedDelayReasonEnum(str, Enum):
    """Controlled justification reasons (required iff classification is justified)."""

    CUSTOMER_REQUEST = "customer_request"
    CUSTOMER_CHANGE_ORDER = "customer_change_order"
    MATERIAL_SUPPLIER_DELAY = "material_supplier_delay"
    FORCE_MAJEURE = "force_majeure"
    UPSTREAM_HOLD = "upstream_hold"
    OTHER = "other"


SELECTABLE_DELAY_REASONS: list[str] = [r.value for r in JustifiedDelayReasonEnum]

File: frontend/src/constants/delayTaxonomy.ts
// Mirror of backend/orm/delay_taxonomy.py — keep the two in lockstep.
//
// 3-state model: WorkOrder.delay_classification is NULL (unclassified,
// default), 'justified', or 'unjustified'. Unclassified is the ABSENCE of
// a value — never an enum member, never offered as a selectable code here
// (the UI still needs a label for it: classificationLabelKey('unclassified')
// resolves fine even though 'unclassified' isn't in
// DELAY_CLASSIFICATION_CODES, same as downtimeTaxonomy's uncategorized
// render key). justified_delay_reason is populated only when
// classification === 'justified'.

export const DELAY_CLASSIFICATION_CODES: string[] = ['justified', 'unjustified']

export const JUSTIFIED_DELAY_REASON_CODES: string[] = [
  'customer_request',
  'customer_change_order',
  'material_supplier_delay',
  'force_majeure',
  'upstream_hold',
  'other',
]

const camel = (v: string): string => {
  const parts = v.toLowerCase().split('_')
  return parts[0] + parts.slice(1).map((p) => p[0].toUpperCase() + p.slice(1)).join('')
}

export const classificationLabelKey = (id: string): string => `delay.classifications.${id}`
export const delayReasonLabelKey = (id: string): string => `delay.reasons.${camel(id)}`

export interface DelayBadgeRow {
  is_late?: boolean
  delay_classification?: string | null
}

export interface DelayBadge {
  key: 'unclassified' | 'justified' | 'unjustified'
  color: 'warning' | 'info' | 'error'
}

// null when the order isn't late (nothing to classify); otherwise
// unclassified/justified/unjustified per the 3-state model above.
export const delayBadge = (row: DelayBadgeRow): DelayBadge | null => {
  if (!row.is_late) return null
  if (row.delay_classification === 'justified') return { key: 'justified', color: 'info' }
  if (row.delay_classification === 'unjustified') return { key: 'unjustified', color: 'error' }
  return { key: 'unclassified', color: 'warning' }
}
